### PR TITLE
feat: osmosis swapper

### PR DESCRIPTION
**Pull request by toshiSat**
## Description

Implemented Osmosis Swapper
Update feeAsset when chain is changed
Made Cosmos chain feeAsset into OSMO

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

 When switching to a new chain an error throws and no quote is available to be grabbed.  I fixed this by auto grabbing an available asset, but product wanted a more verbose error and to not auto select an available token.  This is being handled by the engineering workstream.  EDIT: Fixed

## Risk

None, behind feature flag

## Testing

OSMO -> ATOM swaps will be working now, as well as vice versa.  We can see transactions on mintscan and fees being passed through

### CODE REVIEW

**Comment by elmutt** [comment]
w0000000 there we go

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (elmutt) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:113]: formatting issues?

Comment (toshiSat) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:136]: Idk, I ran `yarn lint --fix` and this happened

**Comment by toshiSat** [comment]
> > 
> 
> Can you explain the bugs you fixed in the trade component?

Yea, so when switching out of eth chain it used to throw an error, I made it to where it auto selects the first available asset (minus the sellasset) 

I made it so it didn't automatically revert to the default assets when switching chains (taking this out since you made a better solution)

Also, I made it so the feeAsset also changes when changing chains, was seeing some funny stuff when trying to get a quote with amount 0 when first changing chains

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (elmutt) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:207]: If this `else if` section needed? wouldn't `sellTradeAsset.asset.assetId` be the same thing so the else would handle it?

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (elmutt) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:202]: Is this needed? would doing nothing work instead?

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:507]: this generic isn't correct here

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:511]: this can come from the swapper display name or similar off the instance

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:209]: this is pretty convoluted - are you trying to get the fee asset from the sell asset? there's a way to get a fee asset from any given asset

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:260]: similarly, this should be like `if (isSupportedSwappingChain(sellAsset.chainId))`

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (toshiSat) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:209]: how do I get the fee asset of ATOM to be Osmosis?

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (toshiSat) [reply src/components/Trade/TradeConfirm/TradeConfirm.tsx:134]: because the explorerTxLink of Atom points to cosmos mintscan and not osmosis.  I mentioned this in a conversation earlier on in the PR

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (elmutt) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:207]: feeAsset here is referring to the fee asset we pay to broadcast the transaction.

Its an interesting problem because now we technically have 2 fee assets because we are broadcasting on 2 chains.

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (elmutt) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:209]: This is a weird problem that the platform doesnt currently handle. see my comment here

https://github.com/shapeshift/web/pull/2128#discussion_r923804193

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (toshiSat) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:207]: hmm yea that naming was confusing then.  We should probably rename it, but I believe that's outside the scope of this ticket

File: sample.env
Comment (toshiSat) [reply sample.env:21]: yea, I can't remember the issue, but it doesn't work at all for me

**Review by elmutt — APPROVED** [review]

**Comment by elmutt** [comment]
LGTM!

**Review by elmutt — APPROVED** [review]
nice. lgtm

File: sample.env
Comment (0xdef1cafe) [reply sample.env:21]: marley pointed me to this directory 

https://cosmos.directory/osmosis/nodes

the https://rpc.osmosis.zone/ node hosted by the foundation is likely to be a little more reliable - assuming the foundation is the most likely node to be stable?

@toshiSat do you have any anecdotal data of using this node?

similarly, we might want to consider using this node for cosmos [https://rpc.cosmos.network:443](https://rpc.cosmos.network/)

Comment (toshiSat) [reply sample.env:21]: I will test these out right now, ty

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:200]: i'm also a bit confused about this logic - in what case do we not have a `sellTradeAsset`?

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:232]: thanks for abstracting this - do you agree that this feels like something the swapper manager should be able to answer? e.g. `swapperManager.isSupportedChain(chainId: ChainId): boolean`

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (toshiSat) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:200]: It looks like it's on initial load maybe?  This was existing code

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (toshiSat) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:232]: Hmm that seems like weird logic.  I could see it returning a list of swappers, but a boolean sounds weird.  Like is it supported across all swappers?  When would we need this?

File: sample.env
Comment (toshiSat) [reply sample.env:21]: Neither of the nodes were the type I needed.  I tried various nodes from that list and certain parts broke.  What is the problem with the nodes I picked? The Osmosis one is listed there.

Comment (0xdef1cafe) [reply sample.env:21]: @toshiSat there's no immediate problem per se - i'm looking at this through the lens of maintenance burden. engineering has to maintain this feature going forward.

ideally we should outsource this to a reputable and accountable third party, e.g. taxistake, where we mutually have skin in the game.

the issue is we'll have potentially an entire chain (osmo), or least features or parts thereof, reliant on an unknown third party, where we have zero insight or recourse with failures, and an unknown recovery time when something goes wrong.

i'm evaluating this in terms of "how reliable and low maintenance will this be going forward, and how do we recover from failures?" vs "does this work right now?"

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (gomesalexandre) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:128]:
```suggestion
fromChainId(trade.sellAsset).chainNamespace === CHAIN_NAMESPACE.COSMOS
```

This will be more future-proof, it will support all IBC chains automagically since they all share the Cosmos chain namespace

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (gomesalexandre) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:232]:
```suggestion
  const isSupportedSwappingChain = (chainId: ChainId) => {
```

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (gomesalexandre) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:260]: typo
```suggestion
    if (!chainAdapter) throw new Error(`Couldn't get chain adapter for ${receiveAddressChainId}`)
```

File: src/components/Trade/hooks/useSwapper/useSwapper.ts
Comment (gomesalexandre) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:282]: If you make `supportedSwappingChain(sellAsset.chainId` a type guard that returns `is | KnownChainIds.EthereumMainnet | KnownChainIds.CosmosMainnet | KnownChainIds.OsmosisMainnet`, `chainId` will be typed properly here and you won't have to do any casting

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx
Comment (toshiSat) [reply src/components/Trade/TradeConfirm/TradeConfirm.tsx]: bug in your suggested change.  should be `trade.sellAsset.chainId`

Comment (toshiSat) [reply src/components/Trade/TradeConfirm/TradeConfirm.tsx]: Also CHAIN_NAMESPACE.Cosmos

File: sample.env
Comment (toshiSat) [reply sample.env:21]: Using proxy nodes ```REACT_APP_OSMOSIS_NODE=https://rest.cosmos.directory/osmosis/
REACT_APP_COSMOS_NODE= https://rest.cosmos.directory/cosmoshub/```

These should stay alive and point to live nodes.
I think these should be good for now.  Still relying on cosmos.directory proxy nodes to be active

**Comment by 0xApotheosis** [comment]
@toshiSat can you please fill out the "Testing" section of the template for your reviewers?

E.g. what should now work? What is an example asset pair that I should expect to be able to test swapping?

File: sample.env
Comment (0xApotheosis) [inline sample.env:21]: Please add these to `pr-cypress.yml` and `pr.yml` too.

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]

File: src/config.ts
Comment (0xdef1cafe) [inline src/config.ts:60]: @toshiSat thanks for changing this - a directory seems more robust here, assuming they're all the same "variety" of nodes.

i know you said a handful of them you tried didn't work - is this the case that they're offline, or that it's similar to infura in that we need an archival node, and not just any kind of node? is there a specific flavor/tier of node that we need, and are we ensuring with these endpoints they're going to provide us those nodes that are guaranteed to work?

would you mind adding a small comment here for maintainers in future?

**Review by 0xdef1cafe — COMMENTED** [review]
small comment as to the source/requirement of nodes that we need for maintainability

**Comment by 0xdef1cafe** [comment]
@toshiSat additionally - i'm not able to see OSMO or ATOM in the ephemeral environment with the flag on.

does this require testing locally with the .env var set?

**Comment by 0xApotheosis** [comment]
Thanks for the testing instructions @toshiSat. When attempting to swap `OSMO` for `ATOM`, or vice versa, I get a `could not find pool` error. Is there something else I need to do to get this to work?


Screen Shot 2022-07-21 at 7 48 04 am

**Comment by toshiSat** [comment]
> Thanks for the testing instructions @toshiSat. When attempting to swap `OSMO` for `ATOM`, or vice versa, I get a `could not find pool` error. Is there something else I need to do to get this to work?
> 
> Screen Shot 2022-07-21 at 7 48 04 am

This is described under issues.  I had this fixed by auto swapping to the next asset, but was told by product to leave it and it was decided to fall under the thorchain work.  Ryan was given the task to show the error message.  Enter in a value and you will get passed this

File: src/config.ts
Comment (toshiSat) [reply src/config.ts:60]: We need a REST node that has CORS enabled.  A lot of them that I tried did not, such as our own hosted datahub nodes

**Comment by toshiSat** [comment]
> @toshiSat additionally - i'm not able to see OSMO or ATOM in the ephemeral environment with the flag on.
> 
> does this require testing locally with the .env var set?

Maybe because i didn't add it to the yml files?  I'm not sure why it's not working on there. I'll test after I add those

**Comment by elmutt** [comment]
> > @toshiSat additionally - i'm not able to see OSMO or ATOM in the ephemeral environment with the flag on.
> > does this require testing locally with the .env var set?
> 
> Maybe because i didn't add it to the yml files? I'm not sure why it's not working on there. I'll test after I add those
> 
> EDIT: Actually those wouldn't have anything to do with it. Hmm I'll look into this now

This is because of how getSwapperManager returns a singleton to the feature flags dont toggle on and off correctly for swappers.

I was planning on fixing this in a followup PR.

Because this is behind a feature flag I feel it is safe to merge and we can followup with any other small fixes.  It would be really nice to finally get this in @0xApotheosis @0xdef1cafe @gomesalexandre @toshiSat

**Review by gomesalexandre — COMMENTED** [review]
Tested on both the ephemeral Fleek branch with the flag on, and locally, both on Keplr and Native:
- On localhost, getting signer errors on both Keplr and Native when trying to swap ATOM for OSMO when confirming the swap
  - `Error [Kind: Error] Failed to retrieve account from signer ` for Native
  - `Unmatched chain id with the offline signer ` for Keplr
- On the ephemeral environment, with the flag on, I can't get past through the swap amount input state. I'm getting CSP errors on `REACT_APP_OSMOSIS_NODE_URL` which seems to indicate these also should be added to the ephemeral env

**Comment by toshiSat** [comment]
> Tested on both the ephemeral Fleek branch with the flag on, and locally, both on Keplr and Native:
> 
> * On localhost, getting signer errors on both Keplr and Native when trying to swap ATOM for OSMO when confirming the swap
>   
>   * `Error [Kind: Error] Failed to retrieve account from signer ` for Native
>   * `Unmatched chain id with the offline signer ` for Keplr
> * On the ephemeral environment, with the flag on, I can't get past through the swap amount input state. I'm getting CSP errors on `REACT_APP_OSMOSIS_NODE_URL` which seems to indicate these also should be added to the ephemeral env

Hmm, I just tried swapping locally ATOM -> OSMO and it's working for me (I am seeing issues with getting the ibc transfer from the new node, i can debug that tomorrow).  Have you imported the `REACT_APP_OSMOSIS_NODE_URL` and `REACT_APP_COSMOS_NODE_URL` into your .env?

**Comment by toshiSat** [comment]
> It might be just me, as @gomesalexandre seemed to have success, but when attempting to swap `OSMO` with `ATOM`, or vice versa, after entering a value, I get a `no receiveAddressAccountSpecifier` error. Is this known or expected?
> 
> Screen Shot 2022-07-21 at 11 24 21 am

Definitely not expected. It looks like that code was pushed to main 5 days ago. This branch is broken right now after I just merged with main. Gonna take a break, but I'm not receiving that error 🤔

File: react-app-rewired/headers/csps/chains/osmosis.ts
Comment (0xApotheosis) [reply react-app-rewired/headers/csps/chains/osmosis.ts]: Addressed in https://github.com/shapeshift/web/pull/2128/commits/38032a8cec94883a096467b536e87b19ca696470.

File: src/config.ts
Comment (elmutt) [reply src/config.ts:65]: @0xdef1cafe How did you determine that a `.directory` domain is more robust?

**Review by 0xApotheosis — COMMENTED** [review]
@toshiSat I'm still getting this guy when trying to swap `ATOM`:`OSMO`:

Screen Shot 2022-07-22 at 7 19 48 am

**Comment by toshiSat** [comment]
> @toshiSat I'm still getting this guy when trying to swap `ATOM`:`OSMO`:
> 
> Screen Shot 2022-07-22 at 7 19 48 am

After debugging with @0xApotheosis we discovered that this issue is surrounding KeepKey as the Osmosis branch has not yet been merged.  I think we can ignore this for now due to being behind a feature flag

**Review by 0xApotheosis — APPROVED** [review]
I didn't have any Cosmos tokens on a non-KeepKey device, so couldn't test a transaction, but I'm now getting rates for `OSMO`<>`ATOM` and vice versa, and the errors noted above have been resolved.

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Tested it out locally with Keplr and I was able to get rates and make a few OSMO -> ATOM swaps.
However, ATOM -> OSMO swaps are still giving me `Unmatched chain id with the offline signer `. 

Also two things I've noted, which might be related:

1. We will need to update the Tx parsing so that Osmosis swaps are shown as actual swaps vs. sends currently. Is that something we plant to tackle in a separate issue? 
2. There are two Txs needed to make said swap. One is a `type_url: /osmosis.gamm.v1beta1.MsgSwapExactAmountIn` and the other is an "IBC transfer" type popping up a few seconds after

As a sanity check, I tried the same swap on osmosis.zone and it only requires a single `type: osmosis/gamm/swap-exact-amount-in` Tx

**Comment by elmutt** [comment]
> Tested it out locally with Keplr and I was able to get rates and make a few OSMO -> ATOM swaps. However, ATOM -> OSMO swaps are still giving me `Unmatched chain id with the offline signer `.
> 
> Also two things I've noted, which might be related:
> 
> 1. We will need to update the Tx parsing so that Osmosis swaps are shown as actual swaps vs. sends currently. Is that something we plant to tackle in a separate issue?
> 
> image
> 
> 2. There are two Txs needed to make said swap. One is a `type_url: /osmosis.gamm.v1beta1.MsgSwapExactAmountIn` and the other is an "IBC transfer" type popping up a few seconds after
> 
> As a sanity check, I tried the same swap on osmosis.zone and it only requires a single `type: osmosis/gamm/swap-exact-amount-in` Tx

1) We are aware of the txd parsing issue for swaps and will handle that in another PR. THis is behind a feature flag and would be nice to get merged as it has been a long-running PR and is safe to merge

2) It requires 2 transactions and so does osmosis zone if you were to do the same thing. We just hide the process and handle it all behind the scenes whereas on osmosis zone they would need to manually ibc deposit first and then an swap tx

**Comment by elmutt** [comment]
> I didn't have any Cosmos tokens on a non-KeepKey device, so couldn't test a transaction, but I'm now getting rates for `OSMO`<>`ATOM` and vice versa, and the errors noted above have been resolved.

Its fine to not test this PR as its behind a feature flag.

All you need to test is that we didnt break anything else when the feature flag is disabled

We would like to keep things moving and get this merged asap and appreciate the understanding that this is behind.

**Comment by gomesalexandre** [comment]
@elmutt If we want to get this merged and tackle the noted issues as follow-ups, that's good for me. I'll proceed to some regression testing with the flag off and stamp it then.

**Comment by elmutt** [comment]
For Testers:

I saw some comment about issues testing with keplr. note that this bounty only requires native and keepkey support

**Review by elmutt — APPROVED** [review]

**Review by gomesalexandre — APPROVED** [review]
Tested both with the flag off and didn't find any regressions with the existing ZRX swapper

🧪

**Comment by 0xApotheosis** [comment]
> For Testers:
> 
> I saw some comment about issues testing with keplr. note that this bounty only requires native and keepkey support

That said @elmutt KeepKey still does not work. Is there a ticket for that?

**Comment by elmutt** [comment]
> > For Testers:
> > I saw some comment about issues testing with keplr. note that this bounty only requires native and keepkey support
> 
> That said @elmutt KeepKey still does not work. Is there a ticket for that?

Its part of the osmosis bounty requirements

**Pull request merged**
